### DISCUSSION TITLE
Proposal: improved constraint support for generics

### DISCUSSION THREAD

**Discussion by chergey** [dotnet/csharplang#653] [General]
1) inherit on single methods constraint from interface  
```
interface IFoo {
  void DoFoo<T>() where T : struct;
}

class Foo : IFoo {
void DoFoo<T>() where T : struct //without specifying constraint, error
{}
}
```

2) constraint for enums and interfaces, e.g.
```
class Foo<T> where T : enum // interface
{}
```
3) multiple constraints, via OR
```
class Foo<T> where T : enum OR struct // or via ||
{}

```

**Comment by Joe4evr** [comment]
1. [That's intentional](https://blogs.msdn.microsoft.com/ericlippert/2012/03/09/why-not-automatically-infer-constraints/).
2. Support for the `enum` constraint is already tracked at #104 (and `delegate` constraint at #103). An interface is no different from a reference type, and that constraint already exists.
3. And how would that even work? What kind of stuff would need to happen at runtime for that to even be possible? What about IntelliSense? Also, enums already count as a `struct`, since it's just a wrapper over some backing fields.

**Comment by HaloFour** [comment]
@Joe4evr

> Also, enums already count as a `struct`

In the world of generics it's actually a little weirder than that.  Just the constraint `System.Enum` would allow for any actual enum type, plus `System.Enum` itself.  But the combination of constraints `System.Enum` and `struct` would force the generic type argument to be an actual enum type:

```cs
// assume the compiler allowed this today, without a special syntax
public class C1<T> where T : System.Enum { }
public class C2<T> where T : System.Enum, struct { }

var c1 = new C1<System.Enum>();  // legal
var c2 = new C2<System.Enum>(); // compiler error
var c22 = new C2<System.DayOfWeek>(); // legal
```

Even stranger is the combination `System.Enum` and `class`, for which the only valid generic type argument is `System.Enum` itself.

**Comment by Joe4evr** [comment]
Okay, but it's still crazy to think "OR constraints" is even remotely in the ballpark for the LDM team.

**Comment by gurustron** [comment]
@Joe4evr 

Still it would be great to have something like 
`where T: (int, double, ...)`

I'm not sure how it can be implemented, but lately i found myself needing something like this, for example, I have next code:

```
public struct MyVal 
{
	public static bool operator !=(MyVal x, MyVal y);
	public static bool operator ==(MyVal x, MyVal y);
	public static implicit operator MyVal(int value);
	public static implicit operator MyVal(long value);
	public static explicit operator int(MyVal value);
	public static explicit operator long(MyVal value);
....... many more operators
}
```

And I'm trying to do something like this:
```
public static IEnumerable<T> CastMyVal<T>(this IEnumerable<MyVal> source)
{
	foreach (T result in source)
		yield return result;
}
```
which obviously gives me compilation error `Cannot convert type 'RedisValue' to 'T'`

It still would be needed to write quite some code but it would be great if it would not be needed to create all specific overloads for CastMyValToInt and stuff.

As for implementation, may be some kind of duck typing is a go.

**Comment by 4creators** [comment]
Real world interop problem which would be solved by modified generic constraints syntax (simplified for clarity):

```C#
[StructLayout(LayoutKind.Explicit)]
public unsafe struct NativeVector<T> where T : byte, ushort, uint, ulong // compile time error
{
    [FieldOffset(0)]
    public fixed T [4]; 
}
```
Another approach which has relaxed syntax - useless in real world code:
```C#
[StructLayout(LayoutKind.Explicit)]
public unsafe struct NativeVector<T> where T : struct
{
    [FieldOffset(0)]
    public fixed T [4];    // compile time error 
}
```
Currently working solution to the problem is a creation of multitude of NativeVector structures one for every underlying numeric type - altogether 10 (byte, sbyte, short, ushort, int, uint, long, ulong, single, double). Firstly it is a bit of unnecessary work, secondly it pollutes type system with 10 new types which otherwise would be replaced by only one. Now let us imagine what happens if I would need to create additional vectors in 1, 2, 3, 5, 6, 7, 8, 9, 10, 11, 12 ..... 1024 dimensions. 

Simple solution which goes outside of the proposal is as follows
```C#
[StructLayout(LayoutKind.Explicit)]
public unsafe struct NativeVector<T, const int N> where T : byte, sbyte, short, ushort, int, uint, long, ulong, float, double 
{
    [FieldOffset(0)]
    public fixed T [N]; 
}
```
In case N is limited to 16 instead of creating 160 different NativeVector structures one could use single generic type - saving is more than obvious. This example is the essence of generic programming.

**Comment by HaloFour** [comment]
IIRC there's no mechanism in the CLR to permit for a type of dynamic size.  The C# language fakes "fixed" buffers by emitting a nested struct of the const size of the total size of the required memory which is calculated at compile time and set via the `Size` property of the `StructLayout` attribute.

The closest thing that C# might be able to get is to emit a different `NativeVector` type per generic arguments, including `N`.  But that would make C# generics a templating engine, which it is not.  It wouldn't be possible to share those as compatible types, nor write methods that could accept such a type without it explicitly declaring `T` and `N` in the callee.

**Comment by 4creators** [comment]
This proposal is very useful for implementation patterns from #749

**Comment by 4creators** [comment]
@HaloFour IMO there is no need for types with dynamic size. The size is determined based on generic parameters T and N at compile time. `N -> constexpr` means that providing size at compile time instance of given type (let's assume it's instance of Tensor&lt;byte, 16&gt;) which is defined in assembly A1 becomes binary compatible with any other instance of type Tensor&lt;byte, 16&gt; from assembly B1 to Bn. Furthermore there is no need to create additional, full type definitions in every assembly but only refer to the defining assembly and generic parameters. There is no need to use templating mechanism from C++ either despite the fact that feature is based on almost identical feature of C++ templates. Currently coreclr is unable to support that kind of binary semantics but we can implement it - support for this feature requires coreclr changes. 

There is no need to emit different types instead one would emit only different parametrized instances of one type at runtime. This would be close to solutions used in dynamic type languages like javascript but if that does the job it's even better bcs someone already implemented feature and tested it in production.

**Comment by HaloFour** [comment]
@4creators 

You'd need that dynamic sizing to make the buffer work.  Fixed buffers don't exist in the CLR; they are a trick of the compiler, accomplished by expanding the type at compile time.  The mechanism used is incompatible with generics because a different nested struct would have to be emitted for every single generic type argument (which can't happen at compile time).  Something like a generic size argument would explode that into a Cartesian product.

The Javascript type system is effectively the polar opposite from the CLR type system.

**Comment by sighoya** [comment]
If Union Types from [Union and Intersection Types](https://github.com/dotnet/csharplang/issues/399) are allowed, then something like:

`
public void(Union{byte i1; sbyte i2; short i3;...} union, ...)={...}
`

would be an alternative, instead of 

`
public void(T t, ...) where T:byte or sbyte or short={...}
`

The advantage would be not to rely on generics.

**Comment by svick** [comment]
@sighoya The meaning of that can be quite different. For example, compare:

```c#
public T M(T x, T y) where T:byte or sbyte or short;
public Union{byte i1; sbyte i2; short i3;} M(Union{byte i1; sbyte i2; short i3;} x, Union{byte i1; sbyte i2; short i3} y);
```

In the first version, `x`, `y` and the return value all have to have the same type. In the second version, the type of each of those can be different. I think that both variants can sometimes be useful, but the use cases are not the same.

**Comment by sighoya** [comment]
@svick 
Yes, for this special case you need parametric polymorphism.

**Comment by gulshan** [comment]
As per my proposed syntax in #399 the constrained generic type code looks like-
```csharp
public T Add<T>(T x, T y) where T : byte | sbyte | short
    => x + y;
```

**Comment by sighoya** [comment]
It is a nice thing to treat T as Union of byte sbyte and short by:
`T: byte | sbyte | short`
But what is the type of a T?
If typeof(T) is Union<byte,sbyte,short> we may get a problem with diagonal types as proposed by @svick:

`
public void fun(T a,T b) where T:Union<Int,String>
{
...
}
`

We can then call:

`
fun(1,2);
fun("hello","")
`

but also:

`
f(1,"hello")
`

Why?,
because typeof(a)==typeof(b)==Union<Int,String>

A way to solve it is to treat typeof(T) to be either Integer or String depending on what we throw in it, even if we specify T to be a Union<String,Int>

It is only an implementation detail, but it is important.

**Comment by sighoya** [comment]
Question: Why is the code in my comment above not syntax colored?

**Comment by gulshan** [comment]
`f(1,"hello")` would not be allowed because `T` has to be same for `a` and `b`. To support it, one can go for something like-
```csharp
public void fun(T1 a, T2 b) where T1 : int | string, T2 : int | string
```
Or if you want to simplify-
```csharp
public void fun(int | string a, int | string b)
```
Use "```csharp" to start code block. And close it with triple ticks to get syntax coloring.

**Comment by sighoya** [comment]
@gulshan 
it depends on the implementation of Unions if f(1,"hello") is allowed
a and b can have the same type in the case typeof(T)==Union<Int,String>
But if type(T)==String or typeof(T)==Int, the as you described it is not possible.

Most Implementations require the latter in order to deny: f(1,"hello") when T:Union<String,Int>
Thanks for syntax coloring advice
